### DISCUSSION TITLE
Custom login process

### DISCUSSION THREAD

**Discussion by sf1450** [carlmontanari/scrapli#274] [Q&A]
Is there a way to create a custom login function?  We have an oddball setup that requires reading in a multiline password prompt (and doing some crypto stuff with it), then generating the password.  The password creation can't be done until first viewing the password prompt.  It looks like I need to use the auth_pypass and on_open method, but I am having a hard time figuring out to make those work together and how to put the multiline password prompt string into a variable.

Any suggestions on how to do this with scrapli without losing other functionality?

**Comment by carlmontanari** [comment]
you can't with paramiko/ssh2/asyncssh but with the other transports you can easily do this by [disabling channel auth](https://carlmontanari.github.io/scrapli/user_guide/advanced_usage/#auth-bypass) and having a custom on open as you mentioned. the wlc example does this exactly. also will have to futz with prompt pattern matching.

if there's a way to make paramiko/ssh2/asyncssh do this defo open to a pr but as they are all expecting "normal" ssh auth im not sure they support anything like that.

**Comment by sf1450** [comment]
Yes, [that](https://github.com/carlmontanari/scrapli/blob/f6b7cb3b5d4a66e816f2d2a7ac31b8af9850bcd7/examples/non_core_device/wlc.py#L11) is where I was stuck.

How would I go about doing item 1 in the below code snippet?

[cls.transport.read?](https://github.com/carlmontanari/scrapli/blob/f6b7cb3b5d4a66e816f2d2a7ac31b8af9850bcd7/scrapli/transport/base/sync_transport.py#L25)


something with channel?

```python
def crypto_on_open(cls):
"""Example `on_open` function for use with crypto prompt"""
#  1) - not sure how to read in the prompt text here...transport.read?  channel.read?
#       do i need to modify the current expected prompt text, let it find that, then somehow read that in?
m = re.match(r"my (example crypto prompt) here", cls.transport.read())  # not sure how to read in data during the login process
if m:
  #  2) process crypto information to generate password
  print("saw the crypto prompt")
  cls.transport.auth_password = my_crypto_function(m.group(1))

#  continue and send the password
cls.transport.write(cls.transport.auth_password)
cls.transport.write(cls.channel.comms_return_char)

```

**Reply by carlmontanari** [reply]
yes you can read from the transport or channel, whichever you want. channel does some basic logging and cleanup. then do whatever you need. reading/writing directly completely bypasses prompt matching. if you want to use `send_x` then you have to care about prompt matching.
